products/iambeta/api.yaml

```diff
@@ -94,11 +94,199 @@ objects:
+
+          Each value must be a [Common Expression Language](https://opensource.google/projects/cel)
+          function that maps an identity provider credential to the normalized attribute specified
+          by the corresponding map key.
+
+          You can use the `assertion` keyword in the expression to access a JSON representation of
+          the authentication credential issued by the provider.
+
+          The maximum length of an attribute mapping expression is 2048 characters. When evaluated,
+          the total size of all mapped attributes must not exceed 8KB.
+
+          For AWS providers, the following rules apply:
+            - If no attribute mapping is defined, the following default mapping applies:
+              ```
+              {
+                "google.subject":"assertion.arn",
+                "attribute.aws_role":
+                  "assertion.arn.contains('assumed-role')"
+                  " ? assertion.arn.extract('{account_arn}assumed-role/')"
+                  "   + 'assumed-role/'"
+                  "   + assertion.arn.extract('assumed-role/{role_name}/')"
+                  " : assertion.arn",
+              }
+              ```
+            - If any custom attribute mappings are defined, they must include a mapping to the
+              `google.subject` attribute.
+
+          For OIDC providers, the following rules apply:
+            - Custom attribute mappings must be defined, and must include a mapping to the
+              `google.subject` attribute. For example, the following maps the `sub` claim of the
+              incoming credential to the `subject` attribute on a Google token.
+              ```
+              {"google.subject": "assertion.sub"}
+              ```
+      - !ruby/object:Api::Type::String
+        name: 'attributeCondition'
+        description: |
+          [A Common Expression Language](https://opensource.google/projects/cel) expression, in
+          plain text, to restrict what otherwise valid authentication credentials issued by the
+          provider should not be accepted.
+
+          The expression must output a boolean representing whether to allow the federation.
+
+          The following keywords may be referenced in the expressions:
+            * `assertion`: JSON representing the authentication credential issued by the provider.
+            * `google`: The Google attributes mapped from the assertion in the `attribute_mappings`.
+            * `attribute`: The custom attributes mapped from the assertion in the `attribute_mappings`.
+
+          The maximum length of the attribute condition expression is 4096 characters. If
+          unspecified, all valid authentication credential are accepted.
+
+          The following example shows how to only allow credentials with a mapped `google.groups`
+          value of `admins`:
+          ```
+          "'admins' in google.groups"
+          ```
+      - !ruby/object:Api::Type::NestedObject
+        name: aws
```

Comment:
It's likely this field and `oicd` are `input: true`, indicating that the value can't be updated. If it _can_, can you add a test that goes from oidc -> aws or vice versa?

Reply:
Good catch!
Switching the type is indead not supported and will throw a http 400 as it is sticking to the old type (aws when changing from aws to oidc and complaining about missing aws account id).
I added the `input:true` to both attributes and still included the test to change type. This will then do a DELETE for the first version and tries to CREATE the second version. However, this fails with http 409 - Requested entity already exists since the resource name still exists (with state DELETED). Should we do something about this?

Reply:
I'm not aware of anything we can do other than fail in that way, so that behaviour seems right to me.

I'd missed this had a deleted state, also, so thanks for pointing that out! We may want to add [this decoder](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/templates/terraform/decoders/storage_hmac_key.go.erb) (renaming to to treat_deleted_state_as_gone.go.erb) which will treat a deleted state as equivalent to the resource not existing. That behaviour is in line with a few other similar resources.

Reply:
I am trying to figure out what such a custom decoder would do. I have added it, but see no changes in any of the tests. The check for DELETED state is for the hmac example is also included in the custom_check_destroy template. That gets generated into `testAccCheckIAMBetaWorkloadIdentityPoolDestroyProducer` which seems to be never executed during any of the tests.

Is there an example test in any of the other resources that I should add to test the delete and which should fail when not adding the custom code to check for DELETED state?

Reply:
Ah- you'll want to add a `test_check_destroy` similar to https://github.com/GoogleCloudPlatform/magic-modules/blob/master/products/storage/terraform.yaml#L139 to affect the tests!

`CheckDestroy` methods call the API directly to ensure the resource is really deleted.